[SK] 사다리 4-4

sunghyun-k opened the pull request:
테스트 가능한 케이스들을 최대한 입력해봤습니다.
다만 랜덤으로 생성될 경우 테스트 할 수 없는 것들은 이니셜라이저를 새로 만들어 테스트 했습니다.

sunghyun-k committed:
사다리 구조체와 연관된 항목 위치 옮김

sunghyun-k committed:
가로대가 연속으로 놓여지지 않았는지 확인하는 테스트케이스를 추가함

sunghyun-k committed:
정렬 기능 테스트 케이스 추가

sunghyun-k committed:
사다리 열 출력 테스트 케이스 추가

sunghyun-k committed:
사다리타기 결과 테스트케이스를 생성함
사다리를 랜덤으로 만들지 않기 위해 사다리에 새로운 이니셜라이저를 추가함
사람 수 오류 및 사다리 높이 오류 테스트 케이스 생성함

sunghyun-k committed:
테스트 케이스에서 공통으로 사용하는 속성을 묶어냄

godrm commented:
(코드 리뷰하기 전에) 테스트하기 좋은 구조, 테스트가 가능한 구현 방식으로 잘 되어 있던가요?

godrm commented on LadderGame/LadderGame/Ladder.swift:
```diff
@@ -1,19 +1,55 @@
-            ladder[row].insert(LadderComponent.empty, at: 0)
-            ladder[row].append(LadderComponent.empty)
+            ladder[row].insert(Component.empty, at: 0)
+            ladder[row].append(Component.empty)
```
타입이 명확하면 `.empty` 로만 작성해도 될 겁니다. 참고하세요.

sunghyun-k replied:
명확한 부분은 점 만 쓰도록 변경하겠습니다.

godrm commented on LadderGame/LadderGame/Ladder.swift:
```diff
@@ -1,19 +1,55 @@
+    /// 테스트를 위해 추가한 이니셜라이저입니다.
+    init(info: [[Component]]) {
+        self.info = info
+    }
```
단지 테스트를 위한 초기화 생성자를 만드는 것은 좋지 않습니다. 
어떤 경우를 테스트하지 못해서 선택하게 된걸까요? 왜 테스트가 불가능하던가요? 다른 방법은 없었던 걸까요?

sunghyun-k replied:
생성된 사다리에서 올바른 결과가 나오는지 확인하는 테스트 케이스였습니다.
사다리 게임 생성 구조가
(플레이어 목록, 높이) -> LadderGame
_LadderGame 초기화 시작_
 (플레이어 수, 높이) -> Ladder 생성 -> LadderGame 사다리 속성에 삽입
_LadderGame 초기화 끝_
이렇게 되고,
사다리 게임 결과값을 확인하는 `LadderGame.result()` 함수를 테스트하기 위해서는 무작위가 아닌 사다리를 LadderGame에 넣어야 했습니다.
초기화 이후에 let으로 선언된 속성을 임의로 바꿀수 있는 방법을 찾아 보았으나 아직까진 방법 파악을 하지 못했습니다. ㅠㅠ

godrm replied:
랜덤 값 자체는 테스트하기 어렵습니다. 그래서 더 작은 단위로 나누고 테스트 가능한 메소드들이 있어야 테스트 가능한 구조가 됩니다. 그렇지 않으면 지금처럼 메소드를 추가하는 기형적인 구조가 되는거죠. 테스트는 설계하고 구현한 기능을 테스트하려는 것이지, 새로운 코드를 추가하면 안됩니다.

sunghyun-k replied:
테스트 케이스를 위한 초기화 함수를 다시 제거하고 `result()` 함수를 수정하여 테스트에 용이하도록 만들겠습니다.

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -1,46 +1,27 @@
     
     //MARK: 속성
     let ladder: Ladder
     let players: [Player]
```
private 으로 선언하지 않은 이유는 무엇인가요? 항상 private로 하고 다른 방법을 찾아보면 좋겠습니다.

sunghyun-k replied:
`LadderGameView`에서 `LadderGame`의 속성을 읽고 출력하도록 되어있어서 공개되어있도록 하였습니다.
private 처리를 위해 `LadderGame`이 출력까지 맡도록 변경을 하는 것이 나을까요?

godrm replied:
아니요. 속성을 그냥 가져가는게 아니라 LadderGame에 메소드를 호출해서 물어봐서 동작하도록 하면 됩니다. 
그 이상 다른 방법들은 다음 스탭에서 연습할 예정입니다.

sunghyun-k replied:
그렇다면 단순히 `LadderGame`안에 `ladder()` 나 `players()`함수를 만들어 각각 값을 반환하도록 하면 된다는 건가요?
만약 제가 이해한 내용이 맞다면 단순히 값을 반환하는 함수를 접근 가능한 속성 대신에 사용하는 이유가 무엇인가요?

godrm replied:
속성으로 하면 가져갔는지 바꿨는지 확인할 방법이 없습니다. 
메소드로 하면 권한을 확인한다거나 다른 값을 준다거나 안준다거나 변경이 가능하죠. 
물론 다음 스탭부터는 그대로 가져가는 것도 하지 않도록 권장할 겁니다.

sunghyun-k replied:
이해가 되었습니다. 함수로 값을 반환하도록 변경하겠습니다.

sunghyun-k commented:
> (코드 리뷰하기 전에) 테스트하기 좋은 구조, 테스트가 가능한 구현 방식으로 잘 되어 있던가요?

오류 처리 시에 throw 구문이 실제 객체의 사용의도와 다른 곳에 배치되어 있어서 테스트 시에 원활하게 테스트 할 수 있도록 자리를 옮기는 과정을 진행했습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -3,11 +3,12 @@ import Foundation
     let playerInfo = try InputControl.readNameOfPlayers()
     let height = try InputControl.readHeight()
-    let ladderGame = LadderGame(players: playerInfo.Players, height: height)
+    let ladderGame = try LadderGame(players: playerInfo.Players, height: height)
```
최상위 함수라면 여기서 아래처럼 catch를 빼기 보다는 각각 try 경우를 예외처리를 하는게 좋습니다. 
(사실 더 갈 곳이 없다고 생각해야 합니다)

sunghyun-k replied:
catch를 빼고 각각의 모든 경우를 잡도록 하겠습니다.

godrm commented on LadderGame/LadderGameTests/LadderGameTests.swift:
```diff
@@ -0,0 +1,145 @@
+    override func setUp() {
+        height = 10
+        numberOfPlayers = 10
+        maxLength = 10
```
setUp() 에는 모든 테스트 케이스 함수에서 다 필요한 경우만 쓰면 됩니다. 
만약 다른 함수에서 다른 값이 필요하면 여기다 쓸 필요가 없다는 얘기입니다.
구체적인 테스트 케이스 더미값들을 테스트 함수 내부에 선언해도 상관없습니다.
테스트 함수들이 독립적으로 동작하기 위해서 고민해야 할 부분 중에 하나입니다.

sunghyun-k replied:
기본적으로 넣었을 때 문제 없는 값들만 setUp()에 넣었습니다.
문제가 발생하지 않는 값들은 따로 뺄 필요 없이 그냥 함수 호출시에 하드코딩을 하는게 일반적인가요?

godrm replied:
setUp() 함수는 모든 테스트 함수가 불릴때 공통적으로 불립니다. 
개별 테스트 케이스는 독립적으로 실행하기 때문에 별도로 작성하는게 좋습니다. 
테스트를 위한 더미값들은 하드코딩을 많이 합니다만, 테스트 함수 내에 테스트를 위한 임시 객체를 만들기도 합니다.

godrm commented on LadderGame/LadderGameTests/LadderGameTests.swift:
```diff
@@ -0,0 +1,145 @@
+        
+    }
+    
+    func testAligning() {
```
테스트 케이스 함수 이름은 호출하려는 용도가 아니라 이름만 보고 어떤 값이 어떤 상황에서 어떻게 처리되야 하는지 구체적으로 알 수 있어야 합니다. 더 자세하게 작성하는 게 좋습니다.

sunghyun-k replied:
테스트 케이스 함수 이름을 구체적으로 작성해보겠습니다.

godrm commented on LadderGame/LadderGameTests/LadderGameTests.swift:
```diff
@@ -0,0 +1,145 @@
+        // Then
+        for number in numberOfPlayers {
+            XCTAssertThrowsError(try Ladder(numberOfPlayers: number, height: height))
+        }
```
여기는 모두다 실패하는 케이스인가요? 
그렇더라도 -5 일때 , 0 일때 1 일때 각각 테스트 케이스를 만드는 게 좋습니다.
구체적으로 어떤 경우가 실패한 것인지 알 수 있도록 구분해보세요.

sunghyun-k replied:
어떤 경우가 실패하는 건지 확인할 생각을 못했네요. 어느 테스트에서 실패했는지 알 수 있도록 구분해보겠습니다.

sunghyun-k committed:
do catch문이 모든 에러 케이스를 확인하도록 변경

sunghyun-k committed:
사다리를 직접 입력받을 수 있도록 result 함수를 수정함.
테스트 케이스를 위한 이니셜라이저 제거
for in 루프로 묶여있던 테스트 케이스에 대해 경우를 나눔

sunghyun-k committed:
사다리 결과 함수 수정
함수에 전달된 매개 변수가 없다면 self의 정보를 사용하도록 변경함

sunghyun-k committed:
테스트 케이스 함수 이름을 보다 명확하게 변경

sunghyun-k committed:
LadderGame의 속성을 비공개로 변경
같은 기능을 하도록 내부 정보를 반환하는 함수를 작성

godrm commented on LadderGame/LadderGameTests/LadderGameTests.swift:
```diff
@@ -84,26 +70,23 @@ class LadderGameTests: XCTestCase {
-        }
+        XCTAssertThrowsError(try Ladder(numberOfPlayers: numberOfPlayers, height: 0))
+        XCTAssertThrowsError(try Ladder(numberOfPlayers: numberOfPlayers, height: -1))
+        XCTAssertThrowsError(try Ladder(numberOfPlayers: numberOfPlayers, height: -5))
```
일반적으로는 Assert도 같은 조건으로 여러개를 사용하지 않는 편입니다. 
저 Assert 중에서 어떤게 실패했는지 명확하게 알수 있기 때문에 테스트 함수는 특정 기능 중에 한 가지 조건만 테스트합니다.
Height 값이 잘못된 경우들 중에서 특별히 한 경우만 테스트하는 거죠

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -34,11 +30,22 @@ struct LadderGame {
+    
+    func gamePlayers() -> [Player] {
+        return players
+    }
```
앞으로는 이렇게 직접 통째로 데이터를 가져가기 보다는 여러 단계로 물어봐서 필요한 데이터를 찾아가는 방식을 고민해보세요.
예를 들어 사다리 세로가 몇 개야? 3개. / 0번째 row는 몇 개야? 5개 / 0, 0 번째 사다리 발판? 0, 1번째 사다리 발판? 이런식으로 상세하게 물어보는 메소드로 구분하는 걸 생각해보라는 겁니다. 
그러면 전체 크기가 얼마나 크거나 작거나 상관없이 접근할 수 있습니다.
지금처럼 만들면 데이터 크기가 커지면 객체를 통째로 복사해야 하는 상황이 생깁니다. 참고하세요.
